src/UriSetHostCommon.c

```diff
@@ -0,0 +1,258 @@
+ */
+
+/* What encodings are enabled? */
+#include <uriparser/UriDefsConfig.h>
+#if (!defined(URI_PASS_ANSI) && !defined(URI_PASS_UNICODE))
+/* Include SELF twice */
+# ifdef URI_ENABLE_ANSI
+#  define URI_PASS_ANSI 1
+#  include "UriSetHostCommon.c"
+#  undef URI_PASS_ANSI
+# endif
+# ifdef URI_ENABLE_UNICODE
+#  define URI_PASS_UNICODE 1
+#  include "UriSetHostCommon.c"
+#  undef URI_PASS_UNICODE
+# endif
+#else
+# ifdef URI_PASS_ANSI
+#  include <uriparser/UriDefsAnsi.h>
+# else
+#  include <uriparser/UriDefsUnicode.h>
+#  include <wchar.h>
+# endif
+
+
+
+#ifndef URI_DOXYGEN
+# include <uriparser/Uri.h>
+# include <uriparser/UriIp4.h>
+# include "UriCommon.h"
+# include "UriMemory.h"
+# include "UriSetHostBase.h"
+# include "UriSetHostCommon.h"
+#endif
+
+
+
+#include <assert.h>
+
+
+
+int URI_FUNC(InternalSetHostMm)(URI_TYPE(Uri) * uri,
+		UriHostType hostType,
+		const URI_CHAR * first,
+		const URI_CHAR * afterLast,
+		UriMemoryManager * memory) {
+	/* Input validation (before making any changes) */
+	if ((uri == NULL) || ((first == NULL) != (afterLast == NULL))) {
+		return URI_ERROR_NULL;
+	}
+
+	URI_CHECK_MEMORY_MANAGER(memory);  /* may return */
+
+	/* The RFC 3986 grammar reads:
+	 *   authority = [ userinfo "@" ] host [ ":" port ]
+	 * So no user info or port without a host. */
+	if (first == NULL) {
+		if (uri->userInfo.first != NULL) {
+			return URI_ERROR_SETHOST_USERINFO_SET;
+		} else if (uri->portText.first != NULL) {
+			return URI_ERROR_SETHOST_PORT_SET;
+		}
+	}
+
+	if (first != NULL) {
+		switch (hostType) {
+			case URI_HOST_TYPE_IP4:
+				if (URI_FUNC(IsWellFormedHostIp4)(first, afterLast) == URI_FALSE) {
+					return URI_ERROR_SYNTAX;
+				}
+				break;
+			case URI_HOST_TYPE_IP6:
+				{
+					const int res = URI_FUNC(IsWellFormedHostIp6Mm)(first, afterLast, memory);
+					assert((res == URI_SUCCESS) || (res == URI_ERROR_SYNTAX) || (res == URI_ERROR_MALLOC));
+					if (res != URI_SUCCESS) {
+						return res;
+					}
+				}
+				break;
+			case URI_HOST_TYPE_IPFUTURE:
+				{
+					const int res = URI_FUNC(IsWellFormedHostIpFutureMm)(first, afterLast, memory);
+					assert((res == URI_SUCCESS) || (res == URI_ERROR_SYNTAX) || (res == URI_ERROR_MALLOC));
+					if (res != URI_SUCCESS) {
+						return res;
+					}
+				}
+				break;
+			case URI_HOST_TYPE_REGNAME:
+				if (URI_FUNC(IsWellFormedHostRegName)(first, afterLast) == URI_FALSE) {
+					return URI_ERROR_SYNTAX;
+				}
+				break;
+			default:
+				assert(0 && "Unsupported URI host type");
+		}
+	}
+
+	{
+		/* Clear old value */
+		const UriBool hadHostBefore = URI_FUNC(HasHost)(uri);
+		if (uri->hostData.ipFuture.first != NULL) {
+			if ((uri->owner == URI_TRUE) && (uri->hostData.ipFuture.first != uri->hostData.ipFuture.afterLast)) {
+				memory->free(memory, (URI_CHAR *)uri->hostData.ipFuture.first);
+			}
+			uri->hostData.ipFuture.first = NULL;
+			uri->hostData.ipFuture.afterLast = NULL;
+			uri->hostText.first = NULL;
+			uri->hostText.afterLast = NULL;
```

Comment:
Should the:
```
			/* NOTE: .hostData.ipFuture holds the very same range pointers
			 *       as .hostText; we must not free memory twice. */
```
comment from UriParse.c be replicated here?

Reply:
> why now…

Only started looking at that file in the last review iteration due to complexity / available time.